### PR TITLE
Change workers rip logics

### PR DESCRIPTION
**Pull request by comtihon**
I changed workers rip algorythm.
I think, that instead of creating timer per worker we should use one timer. It redues messages up to one per OverflowTTL time (now it is one per overflow worker).
Also there are no need of cancelling rip timers. 
Ripped worker ets is also redundant. All rip work is now held on Workers list.

### CODE REVIEW

File: src/poolboy.erl [hunk 1]
```diff
@@ -413,3 +366,23 @@ state_name(State = #state{overflow = Overflow}) when Overflow > 0 ->
...
+    {0, lists:reverse(Workers) ++ UnScanned};
+do_reap_workers([], _, _, Overflow, _, Workers) ->
+    {Overflow, lists:reverse(Workers)};
+do_reap_workers([Worker = {Pid, LTime} | Rest], Now, TTL, Overflow, Sup, Workers) ->
+    case timer:now_diff(Now, LTime) > TTL of
+        true ->
+            ok = dismiss_worker(Sup, Pid),
+            do_reap_workers(Rest, Now, TTL, Overflow - 1, Sup, Workers);
+        false ->
+            do_reap_workers(Rest, Now, TTL, Overflow, Sup, [Worker | Workers])
```
Comment (zugolosian) [inline src/poolboy.erl]: If you've got a really large list of overflow workers and the workers are checked in, in order shouldn't we stop processing the list as soon as we hit one that is false?

File: src/poolboy.erl [hunk 2]
```diff
@@ -248,42 +244,35 @@ handle_info({'EXIT', Pid, _Reason}, State) ->
...
-        [] ->
-            NewState = purge_worker(Pid, State),
-            {noreply, NewState}
-    end;
+handle_info({rip_workers, TimerTTL}, State) ->
+    #state{overflow = Overflow,
+        overflow_ttl = OverTTL,
+        workers = Workers,
+        supervisor = Sup} = State,
+    Now = os:timestamp(),
```
Comment (zugolosian) [inline src/poolboy.erl]: To avoid time warp, it's best to use erlang monotonic so that if the system clock changes underneath us we don't necessarily reap workers. See http://erlang.org/doc/apps/erts/time_correction.html

File: src/poolboy.erl [hunk 3]
```diff
@@ -10,6 +10,7 @@
 -export_type([pool/0]).
 
 -define(TIMEOUT, 5000).
+-define(RIP_PERIOD, 1000).  %1 sec period. If overflow_ttl is less, than a second - rip_period will be same as overflow_ttl
```
Comment (zugolosian) [inline src/poolboy.erl]: Why did you change the terminology to RIP? Reap means to cut or gather.

File: src/poolboy.erl [hunk 4]
```diff
@@ -143,18 +141,19 @@ init([{strategy, lifo} | Rest], WorkerArgs, State) ->
 init([{strategy, fifo} | Rest], WorkerArgs, State) ->
     init(Rest, WorkerArgs, State#state{strategy = fifo});
 init([{overflow_ttl, OverflowTtl} | Rest], WorkerArgs, State) when is_integer(OverflowTtl) ->
-    init(Rest, WorkerArgs, State#state{overflow_ttl = OverflowTtl});
+    TimerTTL = min(OverflowTtl, ?RIP_PERIOD),
```
Comment (zugolosian) [inline src/poolboy.erl]: Can we make this item configurable?

Also can we make the case and naming consistent eg. ReapTime TTL refers to Time To Live the timer that takes care of how frequently we reap workers doesn't have a Time To Live, the overflow workers do.

File: src/poolboy.erl (hunk 3 above)
Comment (comtihon) [inline src/poolboy.erl]: Changed to CHECK_OVERFLOW_DEFAULT_PERIOD

File: src/poolboy.erl (hunk 4 above)
Comment (comtihon) [inline src/poolboy.erl]: I made it configurable and named as overflow_check_period

**Comment by zugolosian** [comment]
@comtihon Did you have any thoughts on the other comments I made on the pull request. I'd like to merge it in once those are addressed.

**Comment by comtihon** [comment]
@zugolosian 
for time wrap - I think it is not good idea to use monotonic time, as it is not necessary provided by all operation systems. I think, that is one decides to change time on servers - there should be much more problems besides.
As for cheking workers while not getting false - we can get situation, when we have old workers at the end of the list, so cheking only for true will get this workers stay.

**Comment by zugolosian** [comment]
Monotonic time as far as I can tell is supported on all operating systems, the guarantees about what it does just change. For os clocks that don't support monotonic it's guaranteed that the time will increase, with os:timestamp it can change either way. 

Consider this, most people using an overflow ttl are using it because starting worker is costly. Keeping workers around a bit longer because their os clock doesn't support monotonicity is better than prematurely killing workers because ntp has (automatically) bought your system time forward.

Can you please explain to me how you would get old workers at the end of the list?

What's your use case? Have you tested the difference in performance between the two branches?

**Comment by comtihon** [comment]
In case of fifo strategy, workers are taken from head and added from tail. So the oldest workers will be in the center of a queue. To gather them we should iterate the whole list.
But in case of lifo (the default strategy) it will be an optimisation, yes.

No I didin't do a performance difference.

Ok, I will use erlang:monotonic_time instead of os:timestamp

**Comment by zugolosian** [comment]
Took a little longer than I expected. 

I will have to spend some time on this and add some tests, will take a look in a few weeks.

**Comment by zugolosian** [comment]
Hi @comtihon 

I've started looking at this again, sorry for the delay. It's mostly looking pretty good. I've created a branch where I've made some minor adjustments. I think there needs to be a test to ensure the timer is working as expected by testing it explicitly.

**Comment by zugolosian** [comment]
Hi @comtihon,

I've changed the implementation a bit in branch https://github.com/zugolosian/poolboy/tree/high_load_issues

Let me know what you think of using that instead?
